src/System.ComponentModel.Composition/tests/Microsoft/Internal/AssumesTests.cs

```diff
@@ -0,0 +1,252 @@
+        }
+
+        [Fact]
+        public void NotNullOfT1T2T3_NullAsValue2ArgumentAndValue3Argument_ShouldThrowInternalErrorException()
+        {
+            Throws(() =>
+            {
+                Assumes.NotNull<string, string, string>("Value", (string)null, (string)null);
+            });
+        }
+
+        [Fact]
+        public void NotNullOfT_ValueAsValueArgument_ShouldNotThrow()
+        {
+            Assumes.NotNull<string>("Value");
+        }
+
+        [Fact]
+        public void NotNullOfT1T2_ValueAsValue1ArgumentAndValue2Argument_ShouldNotThrow()
+        {
+            Assumes.NotNull<string, string>("Value", "Value");
+        }
+
+        [Fact]
+        public void NotNullOfT1T2T3_ValueAsValue1ArgumentAndValue2ArgumentAndValue3Argument_ShouldNotThrow()
+        {
+            Assumes.NotNull<string, string, string>("Value", "Value", "Value");
+        }
+
+        [Fact]
+        public void NotNullOrEmpty_NullAsValueArgument_ShouldThrowInternalErrorException()
+        {
+            Throws(() =>
+            {
+                Assumes.NotNullOrEmpty((string)null);
+            });
+        }
+
+        [Fact]
+        public void NotNullOrEmpty_EmptyAsValueArgument_ShouldThrowInternalErrorException()
+        {
+            Throws(() =>
+            {
+                Assumes.NotNullOrEmpty("");
+            });
+        }
+
+        [Fact]
+        public void NotNullOrEmpty_ValueAsValueArgument_ShouldNotThrow()
+        {
+            var expectations = new List<string>();
+            expectations.Add(" ");
+            expectations.Add("  ");
+            expectations.Add("   ");
+            expectations.Add("Value");
+
+            foreach (var e in expectations)
+            {
+                Assumes.NotNullOrEmpty(e);
+            }
+        }
+
+        [Fact]
+        public void IsTrue1_FalseAsConditionArgument_ShouldThrowInternalErrorException()
+        {
+            Throws(() =>
+            {
+                Assumes.IsTrue(false);
+            });
+        }
+
+        [Fact]
+        public void IsTrue2_FalseAsConditionArgument_ShouldThrowInternalErrorException()
+        {
+            Throws(() =>
+            {
+                Assumes.IsTrue(false, "Message");
+            });
+        }
+
+        [Fact]
+        public void IsTrue1_TrueAsConditionArgument_ShouldNotThrow()
+        {
+            Assumes.IsTrue(true);
+        }
+
+        [Fact]
+        public void IsTrue2_TrueAsConditionArgument_ShouldNotThrow()
+        {
+            Assumes.IsTrue(true, "Message");
+        }
+
+        [Fact]
+        public void NotReachable_ShouldAlwaysThrow()
+        {
+            Throws(() =>
+            {
+                Assumes.NotReachable<object>();
+            });
+        }
+
+        private static void Throws(Action action)
+        {
+            try
+            {
+                action();
+                Assert.False(true);
+            }
+            catch (Exception ex)
```

Comment:
This is going to catch the exception from `Assert.False(true)` (although it doesn't trip the visibility check), which strikes me as.... problematic.  Failures of `Assert.False(exceptionType.IsVisible)` are possibly from two sources: that there was no exception thrown by the action, or that the exception type is not visible, and you can't tell except by debug.  It _passing_ is also possibly from two sources: that the exception from the action was not visible, or the assert returned a not-visible exception.
I think I'd prefer something like the following:

```
private static void ThrowsNotVisibleException(Action action)
{
    var thrown = Assert.ThrowsAny<Exception>(action);
    var exceptionType = thrown.GetType();
    Assert.False(exceptionType.IsVisible);
}
```
... which would immediately make it clear only the action will be throwing an exception, of some unknown type.